clients/src/main/java/org/apache/kafka/clients/NodeApiVersions.java

```diff
@@ -67,30 +75,39 @@ public static NodeApiVersions create(Collection<ApiVersion> overrides) {
         return new NodeApiVersions(apiVersions);
     }
 
-    public NodeApiVersions(Collection<ApiVersion> apiVersions) {
-        this.apiVersions = apiVersions;
-        for (ApiVersion apiVersion : apiVersions) {
-            int apiKeyId = apiVersion.apiKey;
+    public NodeApiVersions(Collection<ApiVersion> nodeApiVersions) {
+        this.nodeApiVersions = nodeApiVersions;
+        for (ApiVersion nodeApiVersion : nodeApiVersions) {
+            int nodeApiKey = nodeApiVersion.apiKey;
             // Newer brokers may support ApiKeys we don't know about, ignore them
-            if (ApiKeys.hasId(apiKeyId)) {
-                short version = Utils.min(ProtoUtils.latestVersion(apiKeyId), apiVersion.maxVersion);
-                if (version >= apiVersion.minVersion && version >= ProtoUtils.oldestVersion(apiKeyId))
-                    usableVersions.put(ApiKeys.forId(apiKeyId), version);
+            if (ApiKeys.hasId(nodeApiKey)) {
+                short v = Utils.min(ProtoUtils.latestVersion(nodeApiKey), nodeApiVersion.maxVersion);
+                if (v < nodeApiVersion.minVersion) {
+                    usableVersions.put(ApiKeys.forId(nodeApiKey), NODE_TOO_NEW);
+                } else if (v < ProtoUtils.oldestVersion(nodeApiKey)) {
+                    usableVersions.put(ApiKeys.forId(nodeApiKey), NODE_TOO_OLD);
+                } else {
+                    usableVersions.put(ApiKeys.forId(nodeApiKey), v);
+                }
             }
         }
     }
 
     /**
-     * Return the most recent version supported by both the client and the server.
+     * Return the most recent version supported by both the node and the local software.
      */
     public short usableVersion(ApiKeys apiKey) {
         Short usableVersion = usableVersions.get(apiKey);
-        if (usableVersion == null) {
-            throw new UnsupportedVersionException("The client cannot send an " +
-                    "API request of type " + apiKey + ", because the " +
-                    "server does not understand any of the versions this client supports.");
-        }
-        return usableVersion;
+        if (usableVersion == API_NOT_ON_NODE)
+            throw new UnsupportedVersionException("The remote node does not support " + apiKey);
```

Comment:
I changed the message to use `broker` for now. It's true that it looks a bit out of place given the class name, but the class is internal and we can rename it later. The message may be seen by users and it's good to make it as clear as possible.